### PR TITLE
all entities, repository and services was implemented

### PR DESCRIPTION
**Pull request by Mandybur800**

### CODE REVIEW

**Review by yurwing — APPROVED** [review]

File: src/main/java/ua/bank/bankservice/model/Account.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @Column(name = "account_number", unique = true, nullable = false)
+    private String accountNumber;
+    @Enumerated(EnumType.STRING)
+    @Column(nullable = false)
+    private Currency currency;
+    @Min(0)
+    private double balance;
```
Comment (boroda4436) [inline src/main/java/ua/bank/bankservice/model/Account.java]: BigDecimal

File: src/main/java/ua/bank/bankservice/model/Transaction.java [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ManyToOne
+    @JoinColumn(name = "account_from", nullable = false)
+    private Account accountFrom;
+    @ManyToOne
+    @JoinColumn(name = "account_to", nullable = false)
+    private Account accountTo;
+    private double amount;
```
Comment (boroda4436) [inline src/main/java/ua/bank/bankservice/model/Transaction.java]: BigDecimal

File: src/main/java/ua/bank/bankservice/util/CurrencyConverter.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+import java.io.InputStreamReader;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import org.springframework.stereotype.Component;
+import ua.bank.bankservice.model.Currency;
+
+@Component
+public class CurrencyConverter {
+    public double convert(Currency fromCurrency, Currency toCurrency, int amount) {
+        String urlStr = "https://api.exchangerate.host/convert?from=" + fromCurrency.name()
```
Comment (boroda4436) [inline src/main/java/ua/bank/bankservice/util/CurrencyConverter.java]: Is it a part of the current HW?

File: src/main/java/ua/bank/bankservice/util/CurrencyConverter.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import org.springframework.stereotype.Component;
+import ua.bank.bankservice.model.Currency;
+
+@Component
+public class CurrencyConverter {
+    public double convert(Currency fromCurrency, Currency toCurrency, int amount) {
```
Comment (boroda4436) [inline src/main/java/ua/bank/bankservice/util/CurrencyConverter.java]:
```suggestion
    public BigDecimal convert(Currency fromCurrency, Currency toCurrency, BigDecimal amount) {
```

File: src/main/java/ua/bank/bankservice/util/CurrencyConverter.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+        String urlStr = "https://api.exchangerate.host/convert?from=" + fromCurrency.name()
+                + "&to=" + toCurrency.name() + "&amount=" + amount;
+        try {
+            URL url = new URL(urlStr);
+            HttpURLConnection request = (HttpURLConnection) url.openConnection();
+            request.connect();
+            JsonElement root = JsonParser.parseReader(new InputStreamReader(
```
Comment (boroda4436) [inline src/main/java/ua/bank/bankservice/util/CurrencyConverter.java]: it is not required to create a new InputStream

Comment (boroda4436) [reply src/main/java/ua/bank/bankservice/util/CurrencyConverter.java]: Let's create a separate class `HttpClient` and path the URL as a param. In this class let's call a method from that class. See example in Thursday's or Friday's lesson.

**Review by boroda4436 — CHANGES_REQUESTED** [review]

File: src/main/java/ua/bank/bankservice/util/CurrencyConverter.java (hunk 5 above)
Comment (boroda4436) [reply src/main/java/ua/bank/bankservice/util/CurrencyConverter.java]: Remove this part from current HW

File: src/main/java/ua/bank/bankservice/controllers/DataInject.java [hunk 6]
```diff
@@ -40,14 +41,14 @@ private void init() {
         firstAccount.setCurrency(Currency.UAH);
         firstAccount.setActive(true);
         firstAccount.setUser(adminUser);
-        firstAccount.setBalance(20000);
+        firstAccount.setBalance(new BigDecimal(20000));
```
Comment (boroda4436) [inline src/main/java/ua/bank/bankservice/controllers/DataInject.java]: isn't this constructor deprecated?

Comment (boroda4436) [inline src/main/java/ua/bank/bankservice/controllers/DataInject.java]:
```suggestion
        firstAccount.setBalance(BigDecimal.valueOf(20000));
```

File: src/main/java/ua/bank/bankservice/service/impl/TransactionServiceImpl.java [hunk 7]
```diff
@@ -38,23 +37,22 @@ public void transfer(String fromAccount, String toAccount, int amount) {
...
         transactionFirst.setAccountTo(accountTo);
         transactionFirst.setDate(LocalDateTime.now());
-        transactionFirst.setAmount(amount);
+        transactionFirst.setAmount(new BigDecimal(amount));
         transactionFirst.setType(Transaction.OperationType.OUTCOMING);
         transactionRepository.save(transactionFirst);
-        accountFrom.setBalance(accountFrom.getBalance() - amount);
-        if (accountFrom.getBalance() < 0) {
+        accountFrom.setBalance(accountFrom.getBalance().subtract(new BigDecimal(amount)));
+        if (accountFrom.getBalance().compareTo(new BigDecimal(0)) < 0) {
```
Comment (boroda4436) [inline src/main/java/ua/bank/bankservice/service/impl/TransactionServiceImpl.java]: BigDecimal.ZERO

**Review by boroda4436 — CHANGES_REQUESTED** [review]

File: src/main/java/ua/bank/bankservice/service/AccountService.java [hunk 8]
```diff
@@ -0,0 +1,14 @@
+package ua.bank.bankservice.service;
+
+import java.util.List;
+import ua.bank.bankservice.model.Account;
+
+public interface AccountService {
+    Account create(Account account);
```
Comment (KseniiaMakarova) [inline src/main/java/ua/bank/bankservice/service/AccountService.java]: how about having a single `save()` method here?
and when we will need to block account - set `active` to false in controller and call `save()` as well?

File: src/main/java/ua/bank/bankservice/repository/TransactionRepository.java [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+import java.util.List;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+import ua.bank.bankservice.model.Transaction;
+
+public interface TransactionRepository extends JpaRepository<Transaction, Long> {
+    @Query(value = "FROM transactions t JOIN FETCH t.accountFrom af JOIN FETCH t.accountTo at "
+            + "WHERE af.accountNumber = ?1 OR at.accountNumber = ?1")
+    List<Transaction> getAllByAccountNumber(String accountNumber, Pageable pageable);
```
Comment (KseniiaMakarova) [inline src/main/java/ua/bank/bankservice/repository/TransactionRepository.java]:
```suggestion
    List<Transaction> getAllByAccount(Account account, Pageable pageable);
```
what do you think?

File: src/main/java/ua/bank/bankservice/service/impl/TransactionServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+
+@Service
+@AllArgsConstructor
+public class TransactionServiceImpl implements TransactionService {
+    private final TransactionRepository transactionRepository;
+    private final AccountService accountService;
+
+    @Override
+    public List<Transaction> getAllByAccount(int page, int size, Account account) {
+        PageRequest date = PageRequest.of(page, size,
```
Comment (KseniiaMakarova) [inline src/main/java/ua/bank/bankservice/service/impl/TransactionServiceImpl.java]: let's have a more informative variable name

File: src/main/java/ua/bank/bankservice/service/impl/TransactionServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+        transactionSecond.setAccountFrom(accountFrom);
+        transactionSecond.setAccountTo(accountTo);
+        transactionSecond.setDate(transactionFirst.getDate());
+        transactionSecond.setAmount(new BigDecimal(amount));
+        transactionSecond.setType(Transaction.OperationType.INCOMING);
+        transactionRepository.save(transactionSecond);
+        accountTo.setBalance(accountTo.getBalance().add(transactionSecond.getAmount()));
+        accountService.create(accountTo);
+    }
+
```
Comment (KseniiaMakarova) [inline src/main/java/ua/bank/bankservice/service/impl/TransactionServiceImpl.java]: redundant empty line

File: src/main/java/ua/bank/bankservice/service/impl/TransactionServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+    @Override
+    public List<Transaction> getAllByAccount(int page, int size, Account account) {
+        PageRequest date = PageRequest.of(page, size,
+                Sort.by("date").descending().and(Sort.by("id")));
+        return transactionRepository.getAllByAccountNumber(account.getAccountNumber(), date);
+    }
+
+    @Override
+    @Transactional
+    public void transfer(String fromAccount, String toAccount, int amount) {
```
Comment (KseniiaMakarova) [inline src/main/java/ua/bank/bankservice/service/impl/TransactionServiceImpl.java]:
```suggestion
    public void transfer(Account fromAccount, Account toAccount, int amount) {
```
maybe let's move this to accountService?

File: pom.xml [hunk 13]
```diff
@@ -23,6 +23,21 @@
...
+            <version>4.5.10</version>
+        </dependency>
+        <dependency>
+            <groupId>javax.validation</groupId>
+            <artifactId>validation-api</artifactId>
+            <version>2.0.1.Final</version>
+        </dependency>
+        <dependency>
+            <groupId>com.google.code.gson</groupId>
+            <artifactId>gson</artifactId>
```
Comment (KseniiaMakarova) [inline pom.xml:38]: did you use this at this point?

File: pom.xml [hunk 14]
```diff
@@ -23,6 +23,21 @@
         </maven.checkstyle.plugin.configLocation>
     </properties>
     <dependencies>
+        <dependency>
+            <groupId>org.apache.httpcomponents</groupId>
+            <artifactId>httpclient</artifactId>
```
Comment (KseniiaMakarova) [inline pom.xml:28]: same

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: pom.xml (hunk 13 above)
Comment (Mandybur800) [reply pom.xml:38]: I use these dependencies in testing for the next step in classes, that I don't push

File: src/main/java/ua/bank/bankservice/service/impl/TransactionServiceImpl.java [hunk 15]
```diff
@@ -23,37 +23,34 @@
...
+        transactionSecond.setAccountFrom(fromAccount);
+        transactionSecond.setAccountTo(toAccount);
         transactionSecond.setDate(transactionFirst.getDate());
         transactionSecond.setAmount(new BigDecimal(amount));
         transactionSecond.setType(Transaction.OperationType.INCOMING);
         transactionRepository.save(transactionSecond);
-        accountTo.setBalance(accountTo.getBalance().add(transactionSecond.getAmount()));
-        accountService.create(accountTo);
+        toAccount.setBalance(toAccount.getBalance().add(transactionSecond.getAmount()));
+        accountService.create(toAccount);
```
Comment (KseniiaMakarova) [inline src/main/java/ua/bank/bankservice/service/impl/TransactionServiceImpl.java]: let's rename the `create()` method - we are not creating Account here

File: src/main/java/ua/bank/bankservice/repository/TransactionRepository.java [hunk 16]
```diff
@@ -4,10 +4,11 @@
 import org.springframework.data.domain.Pageable;
 import org.springframework.data.jpa.repository.JpaRepository;
 import org.springframework.data.jpa.repository.Query;
+import ua.bank.bankservice.model.Account;
 import ua.bank.bankservice.model.Transaction;
 
 public interface TransactionRepository extends JpaRepository<Transaction, Long> {
     @Query(value = "FROM transactions t JOIN FETCH t.accountFrom af JOIN FETCH t.accountTo at "
```
Comment (KseniiaMakarova) [inline src/main/java/ua/bank/bankservice/repository/TransactionRepository.java]: do we really need join fetch here?

**Review by KseniiaMakarova — CHANGES_REQUESTED** [review]

File: src/main/java/ua/bank/bankservice/repository/AccountRepository.java [hunk 17]
```diff
@@ -0,0 +1,14 @@
+package ua.bank.bankservice.repository;
+
+import java.util.List;
+import java.util.Optional;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+import ua.bank.bankservice.model.Account;
+
+public interface AccountRepository extends JpaRepository<Account, Long> {
+    @Query("SELECT a FROM accounts a JOIN FETCH a.user u WHERE u.phoneNumber = :phone")
```
Comment (KseniiaMakarova) [inline src/main/java/ua/bank/bankservice/repository/AccountRepository.java]:
```suggestion
    @Query("SELECT a FROM accounts a WHERE a.user.phoneNumber = :phone")
```

**Review by KseniiaMakarova — APPROVED** [review]

**Pull request merged**